### PR TITLE
[L0][UR] Only Override max allocation limits given env

### PR DESCRIPTION
**Pull request by nrspruit**
pre merge CI PR for https://github.com/oneapi-src/unified-runtime/pull/1245

### CODE REVIEW

**Review by kbenzie — APPROVED** [review]

**Comment by kbenzie** [comment]
Okay, I've managed to reproduce this Window compile error locally. I'll work on a fix to UR and prioritise it.

**Comment by kbenzie** [comment]
An include changed which also includes `Windows.h`, its defining `min`. Fix will be trivial.

**Comment by kbenzie** [comment]
https://github.com/oneapi-src/unified-runtime/pull/1258 should do the trick.

**Comment by kbenzie** [comment]
> [oneapi-src/unified-runtime#1258](https://github.com/oneapi-src/unified-runtime/pull/1258) should do the trick.

Actually this isn't required, we already define `NOMINMAX` in the UR CMake but its not replicated in the PI pluing CMake for L0. If you apply this patch @nrspruit we should be able to merge this sooner:

```diff
diff --git a/sycl/plugins/level_zero/CMakeLists.txt b/sycl/plugins/level_zero/CMakeLists.txt
index 5c24988813c1..19a75b4736bd 100644
--- a/sycl/plugins/level_zero/CMakeLists.txt
+++ b/sycl/plugins/level_zero/CMakeLists.txt
@@ -102,6 +102,10 @@ add_sycl_plugin(level_zero
 if (WIN32)
   # 0x800: Search for the DLL only in the System32 folder
   target_link_options(pi_level_zero PUBLIC /DEPENDENTLOADFLAG:0x800)
+  target_compile_definitions(pi_level_zero PRIVATE
+    WIN32_LEAN_AND_MEAN
+    NOMINMAX
+  )
 endif()

 add_dependencies(pi_level_zero ze-api)
```

**Comment by nrspruit** [comment]
> > [oneapi-src/unified-runtime#1258](https://github.com/oneapi-src/unified-runtime/pull/1258) should do the trick.
> 
> Actually this isn't required, we already define `NOMINMAX` in the UR CMake but its not replicated in the PI plugin CMake for L0. If you apply this patch @nrspruit we should be able to merge this sooner:
> 
> ```diff
> diff --git a/sycl/plugins/level_zero/CMakeLists.txt b/sycl/plugins/level_zero/CMakeLists.txt
> index 5c24988813c1..19a75b4736bd 100644
> --- a/sycl/plugins/level_zero/CMakeLists.txt
> +++ b/sycl/plugins/level_zero/CMakeLists.txt
> @@ -102,6 +102,10 @@ add_sycl_plugin(level_zero
>  if (WIN32)
>    # 0x800: Search for the DLL only in the System32 folder
>    target_link_options(pi_level_zero PUBLIC /DEPENDENTLOADFLAG:0x800)
> +  target_compile_definitions(pi_level_zero PRIVATE
> +    WIN32_LEAN_AND_MEAN
> +    NOMINMAX
> +  )
>  endif()
> 
>  add_dependencies(pi_level_zero ze-api)
> ```
Hello @kbenzie , it looks like this flag is also needed for the opencl plugin. Should this be a larger patch for all plugins?

**Comment by nrspruit** [comment]
> > > [oneapi-src/unified-runtime#1258](https://github.com/oneapi-src/unified-runtime/pull/1258) should do the trick.
> > 
> > 
> > Actually this isn't required, we already define `NOMINMAX` in the UR CMake but its not replicated in the PI plugin CMake for L0. If you apply this patch @nrspruit we should be able to merge this sooner:
> > ```diff
> > diff --git a/sycl/plugins/level_zero/CMakeLists.txt b/sycl/plugins/level_zero/CMakeLists.txt
> > index 5c24988813c1..19a75b4736bd 100644
> > --- a/sycl/plugins/level_zero/CMakeLists.txt
> > +++ b/sycl/plugins/level_zero/CMakeLists.txt
> > @@ -102,6 +102,10 @@ add_sycl_plugin(level_zero
> >  if (WIN32)
> >    # 0x800: Search for the DLL only in the System32 folder
> >    target_link_options(pi_level_zero PUBLIC /DEPENDENTLOADFLAG:0x800)
> > +  target_compile_definitions(pi_level_zero PRIVATE
> > +    WIN32_LEAN_AND_MEAN
> > +    NOMINMAX
> > +  )
> >  endif()
> > 
> >  add_dependencies(pi_level_zero ze-api)
> > ```
> 
> Hello @kbenzie , it looks like this flag is also needed for the opencl plugin. Should this be a larger patch for all plugins?

after updating both plugins opencl and level zero there is a new error because this is set elsewhere:
D:\github\_work\llvm\llvm\src\xpti\include\xpti/xpti_trace_framework.hpp(25): error C2220: the following warning is treated as an error
D:\github\_work\llvm\llvm\src\xpti\include\xpti/xpti_trace_framework.hpp(25): warning C4005: 'NOMINMAX': macro redefinition
D:\github\_work\llvm\llvm\src\xpti\include\xpti/xpti_trace_framework.hpp(25): note: 'NOMINMAX' previously declared on the command line

**Comment by kbenzie** [comment]
> after updating both plugins opencl and level zero there is a new error because this is set elsewhere:
> D:\github_work\llvm\llvm\src\xpti\include\xpti/xpti_trace_framework.hpp(25): error C2220: the following warning is treated as an error
> D:\github_work\llvm\llvm\src\xpti\include\xpti/xpti_trace_framework.hpp(25): warning C4005: 'NOMINMAX': macro redefinition
> D:\github_work\llvm\llvm\src\xpti\include\xpti/xpti_trace_framework.hpp(25): note: 'NOMINMAX' previously declared on the command line

Hmm, perhaps we need to move the `windows.h` include out of `ur_util.hpp` to avoid it being transitively included all over the place. I'll try that out.

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 1]
```diff
@@ -57,13 +57,13 @@ if(SYCL_PI_UR_USE_FETCH_CONTENT)
...
-  #     Merge pull request #1186 from hdelan/device-global-hip
-  #     [HIP] Add support for global variable read write
-  set(UNIFIED_RUNTIME_TAG 79c28d0f0713f58358d5080653d95803fd131749)
+  # commit 5d58871ccce3dd605bdc4245819965cc77441946
+  # Merge: 8007b222 0ebaca6e
+  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
+  # Date:   Wed Jan 17 12:35:19 2024 +0000
+  #     Merge pull request #1245 from nrspruit/enable_relaxed_alloc
+  #     [L0] Only Override max allocation limits given env
+  set(UNIFIED_RUNTIME_TAG 5d58871ccce3dd605bdc4245819965cc77441946)
```
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:66]:
```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/kbenzie/unified-runtime.git")
  set(UNIFIED_RUNTIME_TAG 6036025e127e1460e188861db0ce81de659ee7e6)
```

**Review by kbenzie — COMMENTED** [review]
Okay, I've implemented the change to move the Windows.h include out of ur_util.hpp in https://github.com/oneapi-src/unified-runtime/pull/1260. The suggestion below should let us test if its working as I hope, if that passes then I'll merge the PR and we can use the resulting tag from UR's main branch.

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 1 above)
Comment (nrspruit) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: @kbenzie , I updated this right now so we can see if that fixes the issue, thanks!

Comment (kbenzie) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: sorry, the correct commit is eba3905b1cb3e821401f6b9268358b2684be697d

Comment (nrspruit) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: I Just updated to your latest commit, thanks!

**Comment by kbenzie** [comment]
Since some functions moved out of the `ur_util.hpp` header the `pi_unified_runtime` target now needs to link against the `UnifiedRuntimeCommon` library. This should fix [this link error](https://github.com/intel/llvm/actions/runs/7572988337/job/20624096150?pr=12375).

```diff
diff --git a/sycl/plugins/unified_runtime/CMakeLists.txt b/sycl/plugins/unified_runtime/CMakeLists.txt
index 8d7f2f32b415..b2bb94714c0b 100644
--- a/sycl/plugins/unified_runtime/CMakeLists.txt
+++ b/sycl/plugins/unified_runtime/CMakeLists.txt
@@ -150,6 +150,7 @@ set(UNIFIED_RUNTIME_PLUGIN_ARGS
     Threads::Threads
     UnifiedRuntimeLoader
     UnifiedRuntime-Headers
+    UnifiedRuntimeCommon
   INCLUDE_DIRS
     "${UNIFIED_RUNTIME_SRC_INCLUDE_DIR}"
     "${UNIFIED_RUNTIME_COMMON_INCLUDE_DIR}"
```

**Comment by kbenzie** [comment]
My local build did not have `SYCL_ENABLE_XPTI_TRACING=ON` so they were not failing in this way. I've enabled that now and reproduced the error, I'm not really sure why its there yet but looking into it now.

**Comment by kbenzie** [comment]
I've reverted the CMake changes since they were no logner required after moving the Windows.h include, this should also fix the macro redefinition error.

**Comment by aelovikov-intel** [comment]
I see

```
SYCL :: DeviceLib/string_test.cpp
```

failing in post-commit after this PR on Intel Arc A-series GPU. Please fix ASAP, it's been ignored for two days...

@nrspruit @againull

**Comment by nrspruit** [comment]
> I see
> 
> ```
> SYCL :: DeviceLib/string_test.cpp
> ```
> 
> failing in post-commit after this PR on Intel Arc A-series GPU. Please fix ASAP, it's been ignored for two days...
> 
> @nrspruit @againull

Do you have a link to the error or an error log?

**Comment by aelovikov-intel** [comment]
> Do you have a link to the error or an error log?

https://github.com/intel/llvm/actions/runs/7588321461/job/20670983110

**Pull request merged**